cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/preview/DefaultPreviewRequestQueue.java

```diff
@@ -0,0 +1,134 @@
+ * the License.
+ */
+
+package io.cdap.cdap.internal.app.preview;
+
+import com.google.inject.Inject;
+import io.cdap.cdap.app.preview.PreviewRequest;
+import io.cdap.cdap.app.preview.PreviewRequestQueue;
+import io.cdap.cdap.app.preview.PreviewRequestQueueState;
+import io.cdap.cdap.app.store.preview.PreviewStore;
+import io.cdap.cdap.common.ConflictException;
+import io.cdap.cdap.common.app.RunIds;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.service.Retries;
+import io.cdap.cdap.common.service.RetryStrategies;
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.proto.id.ApplicationId;
+import net.jcip.annotations.ThreadSafe;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Optional;
+import java.util.concurrent.BlockingDeque;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+/**
+ * Thread-safe implementation of {@link PreviewRequestQueue} backed by {@link PreviewStore}.
+ */
+@ThreadSafe
+public class DefaultPreviewRequestQueue implements PreviewRequestQueue {
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultPreviewRequestQueue.class);
+  private final PreviewStore previewStore;
+  private final BlockingDeque<PreviewRequest> requestQueue;
+  private final int capacity;
+  private final long waitTimeOut;
+  private final RetryStrategy retryStrategy;
+
+  @Inject
+  public DefaultPreviewRequestQueue(CConfiguration cConf, PreviewStore previewStore) {
+    this.previewStore = previewStore;
+    this.capacity = cConf.getInt(Constants.Preview.WAITING_QUEUE_CAPACITY, 50);
+    this.waitTimeOut = cConf.getLong(Constants.Preview.WAITING_QUEUE_TIMEOUT_SECONDS, 60);
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, "system.preview.store.update.");
+    this.requestQueue = previewStore.getAllInWaitingState().stream()
+      .filter(r -> isValid(r, waitTimeOut))
+      .collect(Collectors.toCollection(() -> new LinkedBlockingDeque<>(capacity)));
+  }
+
+  @Override
+  public Optional<PreviewRequest> poll(byte[] pollerInfo) {
+    while (true) {
+      PreviewRequest previewRequest = requestQueue.poll();
+      if (previewRequest == null) {
+        return Optional.empty();
+      }
+
+      if (!isValid(previewRequest, waitTimeOut)) {
+        LOG.warn("Preview request wth application id {} is timed out. Ignoring it.",
+                 previewRequest.getProgram().getParent());
+        continue;
+      }
+
+      try {
+        Retries.callWithRetries((Retries.Callable<Void, Exception>) () -> {
+          try {
+            previewStore.setPreviewRequestPollerInfo(previewRequest.getProgram().getParent(), pollerInfo);
+          } catch (ConflictException e) {
+            LOG.debug("Preview application with id {} is not present in WAITING state. Ignoring the preview.",
```

Comment:
When this happen, the preview request is not ignored but getting returned instead. The retry callable should return the previewRequest if it persisted successfully, or `null` if conflict, so that you can break the loop based on the value.

Reply:
Fixed.